src/impl/sink_to_file.cpp

```diff
@@ -204,10 +207,60 @@ namespace soralog {
       if ((end - ptr) < sizeof(Event) or appended
           or std::chrono::steady_clock::now()
                  >= next_flush_.load(std::memory_order_acquire)) {
-        next_flush_.store(std::chrono::steady_clock::now() + latency_,
-                          std::memory_order_release);
-        out_.write(begin, ptr - begin);
-        ptr = begin;
+        std::chrono::seconds attempt_delay{1};
+        auto *data_begin = begin;
+        auto *data_end = ptr;
+
+        while (data_end
+               > data_begin) {  // Loop of trying to write into a log-file
+          next_flush_.store(std::chrono::steady_clock::now() + latency_,
+                            std::memory_order_release);
+          auto pos_before_write = out_.tellp();
+          out_.write(data_begin, data_end - data_begin);
+
+          if (out_.fail() or out_.bad()) {
+            const auto *msg =
+                out_.bad()
+                    ? "Critical I/O error (disk full or hardware failure).\n"
+                    : "Write failed (possibly out of disk space).\n";
+            std::cerr << msg;
+            std::cerr.flush();
+
+            switch (at_fault_) {
+              case AtFaultReactionType::TERMINATE:
+                out_ << "Fatal: " << msg;
+                out_.close();
+                exit(EX_IOERR);
+
+              case AtFaultReactionType::WAITING:
+                std::this_thread::sleep_for(attempt_delay);
+                attempt_delay =
+                    std::min(attempt_delay * 2, std::chrono::seconds(60));
+                if (out_.bad()) {
+                  while (true) {  // Loop of trying to reopen a log-file
+                    std::ofstream out;
+                    out.open(path_, std::ios::app);
+                    if (out.is_open()) {
+                      std::swap(out_, out);
+                      break;  // Leave loop of trying to reopen a log-file
+                    }
+                    std::cerr << "Can't " << (out_.is_open() ? "re-" : "")
+                              << "open log file '" << path_
+                              << "': " << strerror(errno) << ". Waiting for "
+                              << attempt_delay.count() << " seconds...\n";
+                    std::cerr.flush();
+                  }
+                } else {
+                  out_.clear();
+                  std::advance(data_begin, out_.tellp() - pos_before_write);
+                }
+                continue;
+
+              case AtFaultReactionType::DROP_BUFFER:;  // Do nothing
```

Comment:
Should we perhaps write a warning to stderr?

Reply:
https://github.com/qdrvm/soralog/blob/5769bdcf66bace4678f7f5cd95ac799afcdd456c/src/impl/sink_to_file.cpp#L226